File: src/scoring/scorePairs.ts
import { Card } from "../interfaces/card.interface";
import { getCombinations } from "../utils/getCombinations";

/**
 * Scores 2 points for each pair of cards with the same rank.
 */
export const scorePairs = (allCards: Card[]): number => {
  let score = 0;
  const combinations = getCombinations(allCards, 2);
  for (const combo of combinations) {
    if (combo[0].rank === combo[1].rank) {
      score += 2;
    }
  }
  return score;
};

File: crib_calc/scoring/score_pairs.py
from crib_calc.card import Card
from crib_calc.utils.get_combinations import get_combinations


def score_pairs(all_cards: list[Card]) -> int:
    score = 0
    for a, b in get_combinations(all_cards, 2):
        if a.rank == b.rank:
            score += 2
    return score
